File: app/src/main/java/com/popalay/cardme/presentation/base/BasePresenter.java
package com.popalay.cardme.presentation.base;

import com.arellomobile.mvp.MvpPresenter;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * Simplifies getting view. Assumes that we don't care if there are multiple views
 * attached and get the random one.
 */
public abstract class BasePresenter<T extends BaseView> extends MvpPresenter<T> {

    private final CompositeDisposable disposables = new CompositeDisposable();

    public void handleBaseError(Throwable throwable) {
        throwable.printStackTrace();
    }

    protected void addDisposable(Disposable disposable) {
        if (disposable.isDisposed()) return;
        disposables.add(disposable);
    }

    @Override public void onDestroy() {
        super.onDestroy();
        disposables.dispose();
    }
}

File: app/src/main/java/com/popalay/cardme/presentation/base/BasePresenter.kt
package com.popalay.cardme.presentation.base

import com.arellomobile.mvp.MvpPresenter

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

/**
 * Simplifies getting view. Assumes that we don't care if there are multiple views
 * attached and get the random one.
 */
abstract class BasePresenter<T : BaseView> : MvpPresenter<T>() {

    private val disposables = CompositeDisposable()

    fun handleBaseError(throwable: Throwable) = throwable.printStackTrace()

    protected fun addDisposable(disposable: Disposable) {
        if (disposable.isDisposed) return
        disposables.add(disposable)
    }

    override fun onDestroy() {
        super.onDestroy()
        disposables.dispose()
    }
}
